datalad/distribution/get.py

```diff
@@ -309,9 +309,30 @@ def _install_subds_from_flexible_source(ds, sm, **kwargs):
                 res.get('path', None) == dest_path:
             _fixup_submodule_dotgit_setup(ds, sm_path)
 
-            # do fancy update
             lgr.debug("Update cloned subdataset {0} in parent".format(dest_path))
-            ds.repo.update_submodule(sm_path, init=True)
+            section_name = 'submodule.{}'.format(sm['gitmodule_name'])
+            # do not use `git-submodule update --init`, it would make calls
+            # to git-config which will not obey datalad inter-process locks for
+            # modifying .git/config
+            sub = GitRepo(res['path'])
+            # TODO we could simulatanously check out a configured branch
+            # and setup a tracking branch. This should be fine, as the implied
+            # config manipulation is local to the subdataset.
+            sub.call_git(['reset', '--hard', sm['gitshasum']])
```

Comment:
Hmm, so this is a bit more aggressive than the logic in `update_submodule`.  We could end up resetting an unrelated branch.  What about checking out gitshasum and then using the same `get_merge_base` check from `update_submodule`?  (It looks like it could probably be pulled out into a helper fairly cleanly, but perhaps I'm missing some complication.)

Another potential problem here is that gitshasum might not have been pulled in with the clone.  `git submodule update` deals with this by following up with a direct fetch of the commit to try to retrieve it.  We could do the same, but at the very least I think we should translate the failed reset into an error result.

<details>
<summary>demo</summary>

Patch to make cloning a local repo behave like cloning a remote one.

```diff
diff --git a/datalad/core/distributed/clone.py b/datalad/core/distributed/clone.py
index 4b3dd898d..f11f99d6e 100644
--- a/datalad/core/distributed/clone.py
+++ b/datalad/core/distributed/clone.py
@@ -426,7 +426,7 @@ def clone_dataset(
             update=1,
             increment=True)

-        clone_opts = {}
+        clone_opts = {"no_local": True}

         if cand.get('version', None):
             clone_opts['branch'] = cand['version']
```

```sh
#!/bin/sh

set -eu

cd "$(mktemp -d "${TMPDIR:-/tmp}"/dl-XXXXXXX)"

git init a
(
    cd a
    git commit --allow-empty -ma
    git init sub
    (
        cd sub
        git commit --allow-empty -msub
        git checkout master^0
        git commit --allow-empty -mdetached
        git config uploadpack.allowAnySHA1InWant true
    )
    git submodule add ./sub
    git commit -m'add sub'
    git -C sub checkout master
)

git clone a b
(
    cd b
    datalad get sub
)
```

</details>

Reply:
Thanks for the comments and pointers. I pushed https://github.com/datalad/datalad/pull/4853/commits/96e7c8f11591574c0d880c524943838b11f5c2e3 that hopefully addresses them.